modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/gossip/RumorValidator.scala

```diff
@@ -0,0 +1,93 @@
+package org.tessellation.sdk.infrastructure.gossip
+
+import cats.data.NonEmptySet._
+import cats.data.{NonEmptySet, Validated, ValidatedNec}
+import cats.effect.Async
+import cats.syntax.all._
+
+import org.tessellation.ext.cats.conversion.order._
+import org.tessellation.ext.cats.syntax.validated._
+import org.tessellation.ext.crypto._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.ID.Id
+import org.tessellation.schema.gossip._
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.infrastructure.gossip.RumorValidator.RumorValidationErrorOr
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.SignedValidator.SignedValidationError
+import org.tessellation.security.signature.{Signed, SignedValidator}
+
+import derevo.cats.{eqv, show}
+import derevo.derive
+
+trait RumorValidator[F[_]] {
+
+  def validate(har: HashAndRumor): F[RumorValidationErrorOr[HashAndRumor]]
+
+}
+
+object RumorValidator {
+
+  def make[F[_]: Async: KryoSerializer](
+    whitelisting: Option[Set[PeerId]],
+    signedValidator: SignedValidator[F]
+  ): RumorValidator[F] = new RumorValidator[F] {
+
+    def validate(
+      har: HashAndRumor
+    ): F[RumorValidationErrorOr[HashAndRumor]] = har match {
+      case (hash, signedRumor) =>
+        for {
+          hashV <- validateHash(hash, signedRumor)
+          originV = validateOrigin(signedRumor)
+          signatureV <- validateSignature(signedRumor)
+          whitelistingV = validateWhitelisting(signedRumor)
+        } yield
+          hashV
+            .productL(originV)
+            .productL(signatureV)
+            .productL(whitelistingV)
+    }
+
+    def validateHash(hash: Hash, signedRumor: Signed[RumorBinary]): F[RumorValidationErrorOr[HashAndRumor]] =
+      signedRumor.value.hashF.map { calculatedHash =>
+        Validated.condNec(calculatedHash === hash, (hash, signedRumor), InvalidHash(calculatedHash, hash))
+      }
+
+    def validateOrigin(signedRumor: Signed[RumorBinary]): RumorValidationErrorOr[Signed[RumorBinary]] =
+      signedRumor.value match {
+        case _: CommonRumorBinary => signedRumor.validNec[RumorValidationError]
+        case rumor: PeerRumorBinary =>
+          val signers = signedRumor.proofs.map(_.id)
+          Validated.condNec(
+            signers.contains(rumor.origin.toId),
```

Comment:
Contains or equals exactly to that given id? 🤔

Reply:
I don't think it matters much. Even if there are more signatures in a rumor, we're going to care only about the `origin` signature.

Reply:
For sure it doesn't break anything but I wonder what should we allow.

Reply:
There is a method on Signed "isSignedBy" or "isSignedExclusivelyBy". Maybe we can reuse that 🤔

Reply:
@TheMMaciek I could intercept rumors and add my signature. It shouldn't invalidate the original rumor 🤔